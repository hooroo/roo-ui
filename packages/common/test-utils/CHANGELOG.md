# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="0.4.0"></a>
# [0.4.0](https://github.com/hooroo/roo-ui/compare/v0.3.3...v0.4.0) (2018-04-26)


### Features

* add theming ([bee896a](https://github.com/hooroo/roo-ui/commit/bee896a))
