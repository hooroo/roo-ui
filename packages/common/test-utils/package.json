{
  "name": "@roo-ui/test-utils",
  "version": "0.0.0",
  "main": "dist/index.js",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "babel src -d dist --copy-files"
  }
}
