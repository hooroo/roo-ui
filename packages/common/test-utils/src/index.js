export { default as shallowWithTheme } from './shallowWithTheme'; // eslint-disable-line import/prefer-default-export
