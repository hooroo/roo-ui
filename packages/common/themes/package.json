{
  "name": "@roo-ui/themes",
  "version": "0.4.0",
  "main": "dist/index.js",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "babel src -d dist --copy-files"
  }
}
