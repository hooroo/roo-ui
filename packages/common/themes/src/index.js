import qantas from './qantas';

export { qantas }; // eslint-disable-line import/prefer-default-export
