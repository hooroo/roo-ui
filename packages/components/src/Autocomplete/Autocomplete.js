import React from 'react';
import PropTypes from 'prop-types';
import Downshift from 'downshift';
import { themeGet } from 'styled-system';
import { css } from 'styled-components';
import { rem } from 'polished';
import lowerCase from 'lodash/lowerCase';
import omit from 'lodash/omit';

import { Box } from '..';

const Root = Box.extend`
  position: relative;
`;

const Menu = Box.extend`
  position: absolute;
  width: 100%;
  margin-top: -${themeGet('space.3')};
  border-left: ${themeGet('borders.1')} ${themeGet('colors.grey.2')};
  border-right: ${themeGet('borders.1')} ${themeGet('colors.grey.2')};
  z-index: 2;
  max-height: ${rem('230px')};
  overflow-y: scroll;
  -webkit-overflow-scrolling: touch;
`;
Menu.displayName = 'Menu';

const MenuItem = Box.extend`
  background-color: ${themeGet('colors.grey.3')};
  border-bottom: ${themeGet('borders.1')} ${themeGet('colors.grey.2')};
  border-left: ${themeGet('borders.1')} transparent;
  padding: ${themeGet('space.3')} ${themeGet('space.4')};

  ${props =>
    props.highlighted &&
    css`
      background-color: white;
      margin-left: -1px;
      border-left: ${themeGet('borders.2')}
                   ${themeGet('colors.brand.secondary')};
    `};
`;
MenuItem.displayName = 'MenuItem';
MenuItem.defaultProps = {
  blacklist: ['highlighted'],
};

export default class Autocomplete extends React.Component {
  static propTypes = {
    items: PropTypes.oneOfType([PropTypes.func, PropTypes.array]).isRequired,
    filterItems: PropTypes.bool,
    restrictValue: PropTypes.bool,
    selectedItem: PropTypes.string,
    children: PropTypes.oneOfType([PropTypes.node, PropTypes.func]),
  }

  static defaultProps = {
    filterItems: true,
    restrictValue: true,
    selectedItem: '',
    children: null,
  }

  constructor(props) {
    super(props);

    this.state = {
      inputValue: this.props.restrictValue ? undefined : this.props.selectedItem,
    };
  }

  getItems = (inputValue) => {
    if (!this.props.filterItems) {
      return this.props.items;
    }

    return this.props.items.filter(item =>
      !inputValue || lowerCase(item).includes(lowerCase(inputValue)));
  }

  handleStateChange = ({ inputValue }) => {
    if (!this.props.restrictValue) {
      this.setState({ inputValue });
    }
  }

  renderMenu = downshiftProps => (
    <Menu>
      {this.getItems(downshiftProps.inputValue).map((item, index) => (
        <MenuItem
          {...downshiftProps.getItemProps({
            index,
            item,
            key: item,
            highlighted: downshiftProps.highlightedIndex === index,
          })}
        >
          {item}
        </MenuItem>
      ))}
    </Menu>
  );

  render() {
    return (
      <Downshift
        {...omit(this.props, Object.keys(Autocomplete.propTypes))}
        onStateChange={this.handleStateChange}
        selectedItem={this.state.inputValue}
      >
        {downshiftProps => (
          <Root
            {...downshiftProps.getRootProps({ refKey: 'innerRef' })}
          >
            {this.props.children(downshiftProps)}
            {downshiftProps.isOpen && this.renderMenu(downshiftProps)}
          </Root>
        )}
      </Downshift>
    );
  }
}
