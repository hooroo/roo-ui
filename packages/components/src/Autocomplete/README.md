# Autocomplete

<!-- STORY -->

## Installation

```shell
$ yarn add @roo-ui/components
```

## Example

```js
import { Autocomplete, Label, Input } from '@roo-ui/components';

export default (
  <Autocomplete items={['Greyhound', 'Whippet', 'Borzoi']}>
    {({ getLabelProps, getInputProps }) => (
      <Fragment>
        <Label {...getLabelProps()}>Favourite dog</Label>
        <Input {...getInputProps()} />
      </Fragment>
    )}
  </Autocomplete>
);
```

## Properties

| Name                | Description                             | Type    | Default | Required? |
|:--------------------|:----------------------------------------|:--------|:--------|:----------|
| `items`             | array of strings to be autocompleted    | `array` | -       | ✔︎         |
| `filterItems`       | filter items based on input             | `bool`  | true    | -         |
| `restrictInput`     | restrict input to the values in `items` | `bool`  | true    | -         |
| `children`          | function to render label and input      | `func`  | -       | -         |

Refer to [Downshift](https://github.com/paypal/downshift#props) for more props; all props other than the ones listed above will be passed to the `<Downshift />` instance.
