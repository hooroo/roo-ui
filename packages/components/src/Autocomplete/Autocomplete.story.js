import React, { Fragment } from 'react';
import { storiesOf } from '@storybook/react';
import { withDocs } from 'storybook-readme';
import { boolean } from '@storybook/addon-knobs/react';

import { Label, Input, Box } from '..';
import Autocomplete from '.';
import README from './README.md';

const items = [
  'Russo-European Laika',
  'St. John\'s Water Dog',
  'Georgian Shepherd Dog',
  'Miniature Schnauzer',
  'Transylvanian Hound',
  'Leonberger',
  'Majorca Shepherd Dog',
  'Nova Scotia Duck Tolling Retriever',
  'German Roughhaired Pointer',
  'Poodle',
];

storiesOf('Components|Autocomplete', module)
  .addDecorator(withDocs(README))
  .add('default', () => (
    <Box textAlign="left">
      <Autocomplete
        items={items}
        restrictValue={boolean('Restrict value', true)}
        filterItems={boolean('Filter items', true)}
      >
        {({ getLabelProps, getInputProps }) => (
          <Fragment>
            <Label {...getLabelProps()}>Favourite dog</Label>
            <Input {...getInputProps()} />
          </Fragment>
        )}
      </Autocomplete>
    </Box>
  ));
