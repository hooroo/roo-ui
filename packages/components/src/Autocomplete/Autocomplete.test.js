import React, { Fragment } from 'react';
import { qantas as theme } from '@roo-ui/themes';
import { mountWithTheme } from '@roo-ui/test-utils';
import { axe } from 'jest-axe';

import Autocomplete from '.';
import { Label, Input } from '..';

fdescribe('<Autocomplete />', () => {
  let wrapper;
  let props;

  const render = () => mountWithTheme(
    <Autocomplete {...props}>
      {({ getLabelProps, getInputProps }) => (
        <Fragment>
          <Label {...getLabelProps()}>Favourite dog</Label>
          <Input {...getInputProps()} />
        </Fragment>
      )}
    </Autocomplete>,
    theme,
  );

  beforeEach(() => {
    props = {
      items: ['Greyhound', 'Whippet', 'Borzoi'],
    };
    wrapper = render();
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('has no accessibility errors', async () => {
    expect(await axe(wrapper.html())).toHaveNoViolations();
  });

  it('does not render a <Menu />', () => {
    expect(wrapper.find('Menu')).toHaveLength(0);
  });

  it('does not render a <MenuItem />', () => {
    expect(wrapper.find('MenuItem')).toHaveLength(0);
  });

  describe('when open', () => {
    beforeEach(() => {
      props.isOpen = true;
      wrapper = render();
    });

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    // NOTE: waiting on some improvements to Downshift that will land in 2.0.0:
    // https://github.com/paypal/downshift/pull/461
    // it('has no accessibility errors', async () => {
    //   expect(await axe(wrapper.html())).toHaveNoViolations();
    // });

    it('renders a <Menu />', () => {
      expect(wrapper.find('Menu')).toHaveLength(1);
    });

    it('renders a <MenuItem /> for each item', () => {
      expect(wrapper.find('MenuItem')).toHaveLength(3);
    });

    it('renders a <MenuItem /> for each item', () => {
      expect(wrapper.find('MenuItem')).toHaveLength(3);
    });
  });

  describe('when filterItems is set', () => {
    beforeEach(() => {
      props.filterItems = true;
      props.inputValue = 'i';
      props.isOpen = true;
      wrapper = render();
    });

    it('filters items by input value', () => {
      expect(wrapper.find('MenuItem')).toHaveLength(2);
      expect(wrapper.find('MenuItem').at(0).text()).toBe('Whippet');
      expect(wrapper.find('MenuItem').at(1).text()).toBe('Borzoi');
    });
  });

  describe('when filterItems is not set', () => {
    beforeEach(() => {
      props.filterItems = false;
      props.inputValue = 'i';
      props.isOpen = true;
      wrapper = render();
    });

    it('does not filter item', () => {
      expect(wrapper.find('MenuItem')).toHaveLength(3);
    });
  });

  describe('when restrictValue is set', () => {
    beforeEach(() => {
      props.restrictValue = true;
      wrapper = render();
      wrapper.find('input').simulate('change', { target: { value: 'Longboy' } });
      wrapper.find('input').simulate('keydown', { key: 'Escape' });
    });

    it('restricts input value to items', () => {
      expect(wrapper.find('input').instance().value).toBe('');
    });
  });

  describe('when restrictValue is not set', () => {
    beforeEach(() => {
      props.restrictValue = false;
      wrapper = render();
      wrapper.find('input').simulate('change', { target: { value: 'Longboy' } });
      wrapper.find('input').simulate('keydown', { key: 'Escape' });
    });

    it('does not restrict input value to items', () => {
      expect(wrapper.find('input').instance().value).toBe('Longboy');
    });
  });
});
