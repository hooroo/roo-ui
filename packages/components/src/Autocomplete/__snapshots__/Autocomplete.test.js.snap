// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Autocomplete /> renders correctly 1`] = `
.c0 {
  position: relative;
}

.c2 {
  display: block;
  width: 100%;
  margin: 0;
  outline: 0;
  -webkit-transition: border-color;
  transition: border-color;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c2[readonly] {
  border-color: transparent;
  background: transparent;
  padding-left: 0;
}

.c2::-ms-clear {
  display: none;
}

.c2::-ms-clear {
  display: none;
}

.c2::-ms-clear {
  display: none;
}

.c2::-ms-clear {
  display: none;
}

.c1 {
  display: block;
  width: 100%;
  margin: 0;
}

<Autocomplete
  filterItems={true}
  items={
    Array [
      "Greyhound",
      "Whippet",
      "Borzoi",
    ]
  }
  restrictValue={true}
  selectedItem=""
>
  <Downshift
    breakingChanges={Object {}}
    defaultHighlightedIndex={null}
    defaultInputValue=""
    defaultIsOpen={false}
    defaultSelectedItem={null}
    environment={[Window]}
    getA11yStatusMessage={[Function]}
    itemToString={[Function]}
    onChange={[Function]}
    onInputValueChange={[Function]}
    onOuterClick={[Function]}
    onSelect={[Function]}
    onStateChange={[Function]}
    onUserAction={[Function]}
    selectedItemChanged={[Function]}
    stateReducer={[Function]}
  >
    <Box
      innerRef={[Function]}
    >
      <Clean.div
        blacklist={
          Array [
            "m",
            "mt",
            "mr",
            "mb",
            "ml",
            "mx",
            "my",
            "p",
            "pt",
            "pr",
            "pb",
            "pl",
            "px",
            "py",
            "width",
            "w",
            "fontSize",
            "f",
            "color",
            "bg",
            "fontFamily",
            "font",
            "textAlign",
            "align",
            "lineHeight",
            "fontWeight",
            "letterSpacing",
            "display",
            "maxWidth",
            "minWidth",
            "height",
            "maxHeight",
            "minHeight",
            "size",
            "ratio",
            "alignItems",
            "alignContent",
            "justifyContent",
            "justify",
            "flexWrap",
            "wrap",
            "flexBasis",
            "flexDirection",
            "flex",
            "justifySelf",
            "alignSelf",
            "order",
            "gridGap",
            "gridColumnGap",
            "gridRowGap",
            "gridColumn",
            "gridRow",
            "gridAutoFlow",
            "gridAutoColumns",
            "gridAutoRows",
            "gridTemplateColumns",
            "gridTemplateRows",
            "border",
            "borderTop",
            "borderRight",
            "borderBottom",
            "borderLeft",
            "borderColor",
            "borderRadius",
            "boxShadow",
            "background",
            "backgroundImage",
            "bgImage",
            "backgroundSize",
            "bgSize",
            "backgroundPosition",
            "bgPosition",
            "backgroundRepeat",
            "bgRepeat",
            "position",
            "zIndex",
            "top",
            "right",
            "bottom",
            "left",
            "hover",
            "focus",
            "active",
            "disabledStyle",
            "textStyle",
            "colors",
            "buttonStyle",
            "borderWidth",
          ]
        }
        className="c0"
        innerRef={[Function]}
        is="div"
      >
        <div
          className="c0"
        >
          <Label
            hidden={false}
            htmlFor="downshift-0-input"
            textStyle="label"
          >
            <Clean.label
              className="c1"
              hidden={false}
              htmlFor="downshift-0-input"
              textStyle="label"
            >
              <Clean.div
                blacklist={
                  Array [
                    "m",
                    "mt",
                    "mr",
                    "mb",
                    "ml",
                    "mx",
                    "my",
                    "p",
                    "pt",
                    "pr",
                    "pb",
                    "pl",
                    "px",
                    "py",
                    "width",
                    "w",
                    "fontSize",
                    "f",
                    "color",
                    "bg",
                    "fontFamily",
                    "font",
                    "textAlign",
                    "align",
                    "lineHeight",
                    "fontWeight",
                    "letterSpacing",
                    "display",
                    "maxWidth",
                    "minWidth",
                    "height",
                    "maxHeight",
                    "minHeight",
                    "size",
                    "ratio",
                    "alignItems",
                    "alignContent",
                    "justifyContent",
                    "justify",
                    "flexWrap",
                    "wrap",
                    "flexBasis",
                    "flexDirection",
                    "flex",
                    "justifySelf",
                    "alignSelf",
                    "order",
                    "gridGap",
                    "gridColumnGap",
                    "gridRowGap",
                    "gridColumn",
                    "gridRow",
                    "gridAutoFlow",
                    "gridAutoColumns",
                    "gridAutoRows",
                    "gridTemplateColumns",
                    "gridTemplateRows",
                    "border",
                    "borderTop",
                    "borderRight",
                    "borderBottom",
                    "borderLeft",
                    "borderColor",
                    "borderRadius",
                    "boxShadow",
                    "background",
                    "backgroundImage",
                    "bgImage",
                    "backgroundSize",
                    "bgSize",
                    "backgroundPosition",
                    "bgPosition",
                    "backgroundRepeat",
                    "bgRepeat",
                    "position",
                    "zIndex",
                    "top",
                    "right",
                    "bottom",
                    "left",
                    "hover",
                    "focus",
                    "active",
                    "disabledStyle",
                    "textStyle",
                    "colors",
                    "buttonStyle",
                    "borderWidth",
                  ]
                }
                className="c1"
                hidden={false}
                htmlFor="downshift-0-input"
                is="label"
                textStyle="label"
              >
                <label
                  className="c1"
                  hidden={false}
                  htmlFor="downshift-0-input"
                >
                  Favourite dog
                </label>
              </Clean.div>
            </Clean.label>
          </Label>
          <Input
            aria-activedescendant={null}
            aria-autocomplete="list"
            aria-expanded={false}
            autoComplete="off"
            blacklist={
              Array [
                "m",
                "mt",
                "mr",
                "mb",
                "ml",
                "mx",
                "my",
                "p",
                "pt",
                "pr",
                "pb",
                "pl",
                "px",
                "py",
                "error",
                "underline",
              ]
            }
            id="downshift-0-input"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyDown={[Function]}
            role="combobox"
            value=""
          >
            <Clean.input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-expanded={false}
              autoComplete="off"
              blacklist={
                Array [
                  "m",
                  "mt",
                  "mr",
                  "mb",
                  "ml",
                  "mx",
                  "my",
                  "p",
                  "pt",
                  "pr",
                  "pb",
                  "pl",
                  "px",
                  "py",
                  "error",
                  "underline",
                ]
              }
              className="c2"
              id="downshift-0-input"
              onBlur={[Function]}
              onChange={[Function]}
              onKeyDown={[Function]}
              role="combobox"
              value=""
            >
              <Clean.div
                aria-activedescendant={null}
                aria-autocomplete="list"
                aria-expanded={false}
                autoComplete="off"
                blacklist={
                  Array [
                    "m",
                    "mt",
                    "mr",
                    "mb",
                    "ml",
                    "mx",
                    "my",
                    "p",
                    "pt",
                    "pr",
                    "pb",
                    "pl",
                    "px",
                    "py",
                    "error",
                    "underline",
                  ]
                }
                className="c2"
                id="downshift-0-input"
                is="input"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyDown={[Function]}
                role="combobox"
                value=""
              >
                <input
                  aria-activedescendant={null}
                  aria-autocomplete="list"
                  aria-expanded={false}
                  autoComplete="off"
                  className="c2"
                  id="downshift-0-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onKeyDown={[Function]}
                  role="combobox"
                  value=""
                />
              </Clean.div>
            </Clean.input>
          </Input>
        </div>
      </Clean.div>
    </Box>
  </Downshift>
</Autocomplete>
`;

exports[`<Autocomplete /> when open renders correctly 1`] = `
.c0 {
  position: relative;
}

.c3 {
  position: absolute;
  width: 100%;
  margin-top: -;
}

.c4 {
  border-left: transparent;
}

.c2 {
  display: block;
  width: 100%;
  margin: 0;
  outline: 0;
  -webkit-transition: border-color;
  transition: border-color;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c2[readonly] {
  border-color: transparent;
  background: transparent;
  padding-left: 0;
}

.c2::-ms-clear {
  display: none;
}

.c2::-ms-clear {
  display: none;
}

.c2::-ms-clear {
  display: none;
}

.c2::-ms-clear {
  display: none;
}

.c1 {
  display: block;
  width: 100%;
  margin: 0;
}

<Autocomplete
  filterItems={true}
  isOpen={true}
  items={
    Array [
      "Greyhound",
      "Whippet",
      "Borzoi",
    ]
  }
  restrictValue={true}
  selectedItem=""
>
  <Downshift
    breakingChanges={Object {}}
    defaultHighlightedIndex={null}
    defaultInputValue=""
    defaultIsOpen={false}
    defaultSelectedItem={null}
    environment={[Window]}
    getA11yStatusMessage={[Function]}
    isOpen={true}
    itemToString={[Function]}
    onChange={[Function]}
    onInputValueChange={[Function]}
    onOuterClick={[Function]}
    onSelect={[Function]}
    onStateChange={[Function]}
    onUserAction={[Function]}
    selectedItemChanged={[Function]}
    stateReducer={[Function]}
  >
    <Box
      innerRef={[Function]}
    >
      <Clean.div
        blacklist={
          Array [
            "m",
            "mt",
            "mr",
            "mb",
            "ml",
            "mx",
            "my",
            "p",
            "pt",
            "pr",
            "pb",
            "pl",
            "px",
            "py",
            "width",
            "w",
            "fontSize",
            "f",
            "color",
            "bg",
            "fontFamily",
            "font",
            "textAlign",
            "align",
            "lineHeight",
            "fontWeight",
            "letterSpacing",
            "display",
            "maxWidth",
            "minWidth",
            "height",
            "maxHeight",
            "minHeight",
            "size",
            "ratio",
            "alignItems",
            "alignContent",
            "justifyContent",
            "justify",
            "flexWrap",
            "wrap",
            "flexBasis",
            "flexDirection",
            "flex",
            "justifySelf",
            "alignSelf",
            "order",
            "gridGap",
            "gridColumnGap",
            "gridRowGap",
            "gridColumn",
            "gridRow",
            "gridAutoFlow",
            "gridAutoColumns",
            "gridAutoRows",
            "gridTemplateColumns",
            "gridTemplateRows",
            "border",
            "borderTop",
            "borderRight",
            "borderBottom",
            "borderLeft",
            "borderColor",
            "borderRadius",
            "boxShadow",
            "background",
            "backgroundImage",
            "bgImage",
            "backgroundSize",
            "bgSize",
            "backgroundPosition",
            "bgPosition",
            "backgroundRepeat",
            "bgRepeat",
            "position",
            "zIndex",
            "top",
            "right",
            "bottom",
            "left",
            "hover",
            "focus",
            "active",
            "disabledStyle",
            "textStyle",
            "colors",
            "buttonStyle",
            "borderWidth",
          ]
        }
        className="c0"
        innerRef={[Function]}
        is="div"
      >
        <div
          className="c0"
        >
          <Label
            hidden={false}
            htmlFor="downshift-5-input"
            textStyle="label"
          >
            <Clean.label
              className="c1"
              hidden={false}
              htmlFor="downshift-5-input"
              textStyle="label"
            >
              <Clean.div
                blacklist={
                  Array [
                    "m",
                    "mt",
                    "mr",
                    "mb",
                    "ml",
                    "mx",
                    "my",
                    "p",
                    "pt",
                    "pr",
                    "pb",
                    "pl",
                    "px",
                    "py",
                    "width",
                    "w",
                    "fontSize",
                    "f",
                    "color",
                    "bg",
                    "fontFamily",
                    "font",
                    "textAlign",
                    "align",
                    "lineHeight",
                    "fontWeight",
                    "letterSpacing",
                    "display",
                    "maxWidth",
                    "minWidth",
                    "height",
                    "maxHeight",
                    "minHeight",
                    "size",
                    "ratio",
                    "alignItems",
                    "alignContent",
                    "justifyContent",
                    "justify",
                    "flexWrap",
                    "wrap",
                    "flexBasis",
                    "flexDirection",
                    "flex",
                    "justifySelf",
                    "alignSelf",
                    "order",
                    "gridGap",
                    "gridColumnGap",
                    "gridRowGap",
                    "gridColumn",
                    "gridRow",
                    "gridAutoFlow",
                    "gridAutoColumns",
                    "gridAutoRows",
                    "gridTemplateColumns",
                    "gridTemplateRows",
                    "border",
                    "borderTop",
                    "borderRight",
                    "borderBottom",
                    "borderLeft",
                    "borderColor",
                    "borderRadius",
                    "boxShadow",
                    "background",
                    "backgroundImage",
                    "bgImage",
                    "backgroundSize",
                    "bgSize",
                    "backgroundPosition",
                    "bgPosition",
                    "backgroundRepeat",
                    "bgRepeat",
                    "position",
                    "zIndex",
                    "top",
                    "right",
                    "bottom",
                    "left",
                    "hover",
                    "focus",
                    "active",
                    "disabledStyle",
                    "textStyle",
                    "colors",
                    "buttonStyle",
                    "borderWidth",
                  ]
                }
                className="c1"
                hidden={false}
                htmlFor="downshift-5-input"
                is="label"
                textStyle="label"
              >
                <label
                  className="c1"
                  hidden={false}
                  htmlFor="downshift-5-input"
                >
                  Favourite dog
                </label>
              </Clean.div>
            </Clean.label>
          </Label>
          <Input
            aria-activedescendant={null}
            aria-autocomplete="list"
            aria-expanded={true}
            autoComplete="off"
            blacklist={
              Array [
                "m",
                "mt",
                "mr",
                "mb",
                "ml",
                "mx",
                "my",
                "p",
                "pt",
                "pr",
                "pb",
                "pl",
                "px",
                "py",
                "error",
                "underline",
              ]
            }
            id="downshift-5-input"
            onBlur={[Function]}
            onChange={[Function]}
            onKeyDown={[Function]}
            role="combobox"
            value=""
          >
            <Clean.input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-expanded={true}
              autoComplete="off"
              blacklist={
                Array [
                  "m",
                  "mt",
                  "mr",
                  "mb",
                  "ml",
                  "mx",
                  "my",
                  "p",
                  "pt",
                  "pr",
                  "pb",
                  "pl",
                  "px",
                  "py",
                  "error",
                  "underline",
                ]
              }
              className="c2"
              id="downshift-5-input"
              onBlur={[Function]}
              onChange={[Function]}
              onKeyDown={[Function]}
              role="combobox"
              value=""
            >
              <Clean.div
                aria-activedescendant={null}
                aria-autocomplete="list"
                aria-expanded={true}
                autoComplete="off"
                blacklist={
                  Array [
                    "m",
                    "mt",
                    "mr",
                    "mb",
                    "ml",
                    "mx",
                    "my",
                    "p",
                    "pt",
                    "pr",
                    "pb",
                    "pl",
                    "px",
                    "py",
                    "error",
                    "underline",
                  ]
                }
                className="c2"
                id="downshift-5-input"
                is="input"
                onBlur={[Function]}
                onChange={[Function]}
                onKeyDown={[Function]}
                role="combobox"
                value=""
              >
                <input
                  aria-activedescendant={null}
                  aria-autocomplete="list"
                  aria-expanded={true}
                  autoComplete="off"
                  className="c2"
                  id="downshift-5-input"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onKeyDown={[Function]}
                  role="combobox"
                  value=""
                />
              </Clean.div>
            </Clean.input>
          </Input>
          <Menu>
            <Clean.div
              blacklist={
                Array [
                  "m",
                  "mt",
                  "mr",
                  "mb",
                  "ml",
                  "mx",
                  "my",
                  "p",
                  "pt",
                  "pr",
                  "pb",
                  "pl",
                  "px",
                  "py",
                  "width",
                  "w",
                  "fontSize",
                  "f",
                  "color",
                  "bg",
                  "fontFamily",
                  "font",
                  "textAlign",
                  "align",
                  "lineHeight",
                  "fontWeight",
                  "letterSpacing",
                  "display",
                  "maxWidth",
                  "minWidth",
                  "height",
                  "maxHeight",
                  "minHeight",
                  "size",
                  "ratio",
                  "alignItems",
                  "alignContent",
                  "justifyContent",
                  "justify",
                  "flexWrap",
                  "wrap",
                  "flexBasis",
                  "flexDirection",
                  "flex",
                  "justifySelf",
                  "alignSelf",
                  "order",
                  "gridGap",
                  "gridColumnGap",
                  "gridRowGap",
                  "gridColumn",
                  "gridRow",
                  "gridAutoFlow",
                  "gridAutoColumns",
                  "gridAutoRows",
                  "gridTemplateColumns",
                  "gridTemplateRows",
                  "border",
                  "borderTop",
                  "borderRight",
                  "borderBottom",
                  "borderLeft",
                  "borderColor",
                  "borderRadius",
                  "boxShadow",
                  "background",
                  "backgroundImage",
                  "bgImage",
                  "backgroundSize",
                  "bgSize",
                  "backgroundPosition",
                  "bgPosition",
                  "backgroundRepeat",
                  "bgRepeat",
                  "position",
                  "zIndex",
                  "top",
                  "right",
                  "bottom",
                  "left",
                  "hover",
                  "focus",
                  "active",
                  "disabledStyle",
                  "textStyle",
                  "colors",
                  "buttonStyle",
                  "borderWidth",
                ]
              }
              className="c3"
              is="div"
            >
              <div
                className="c3"
              >
                <MenuItem
                  blacklist={
                    Array [
                      "highlighted",
                    ]
                  }
                  highlighted={false}
                  id="downshift-5-item-0"
                  key="Greyhound"
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                >
                  <Clean.div
                    blacklist={
                      Array [
                        "highlighted",
                      ]
                    }
                    className="c4"
                    highlighted={false}
                    id="downshift-5-item-0"
                    is="div"
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseMove={[Function]}
                  >
                    <div
                      className="c4"
                      id="downshift-5-item-0"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseMove={[Function]}
                    >
                      Greyhound
                    </div>
                  </Clean.div>
                </MenuItem>
                <MenuItem
                  blacklist={
                    Array [
                      "highlighted",
                    ]
                  }
                  highlighted={false}
                  id="downshift-5-item-1"
                  key="Whippet"
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                >
                  <Clean.div
                    blacklist={
                      Array [
                        "highlighted",
                      ]
                    }
                    className="c4"
                    highlighted={false}
                    id="downshift-5-item-1"
                    is="div"
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseMove={[Function]}
                  >
                    <div
                      className="c4"
                      id="downshift-5-item-1"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseMove={[Function]}
                    >
                      Whippet
                    </div>
                  </Clean.div>
                </MenuItem>
                <MenuItem
                  blacklist={
                    Array [
                      "highlighted",
                    ]
                  }
                  highlighted={false}
                  id="downshift-5-item-2"
                  key="Borzoi"
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                >
                  <Clean.div
                    blacklist={
                      Array [
                        "highlighted",
                      ]
                    }
                    className="c4"
                    highlighted={false}
                    id="downshift-5-item-2"
                    is="div"
                    onClick={[Function]}
                    onMouseDown={[Function]}
                    onMouseMove={[Function]}
                  >
                    <div
                      className="c4"
                      id="downshift-5-item-2"
                      onClick={[Function]}
                      onMouseDown={[Function]}
                      onMouseMove={[Function]}
                    >
                      Borzoi
                    </div>
                  </Clean.div>
                </MenuItem>
              </div>
            </Clean.div>
          </Menu>
        </div>
      </Clean.div>
    </Box>
  </Downshift>
</Autocomplete>
`;
